contracts/HenkakuQuest.sol

```diff
@@ -2,25 +2,144 @@
+        );
+        _grantRole(ADMIN_ROLE, _to);
+    }
+
+    function addQuestCreationRole(address _to) public {
+        require(
+            hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have an admin role to add quest creation role"
+        );
+        _grantRole(QUEST_CREATION_ROLE, _to);
+    }
+
+    function addMemberRole(address _to) public {
+        require(
+            hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have an admin role to add quest creation role"
+        );
+        _grantRole(HENKAKU_MEMBER_ROLE, _to);
+    }
+
+    function isMember(address _address) internal view returns (bool) {
+        uint256 _balance = memberShipNFT.balanceOf(_address);
+        if (_balance <= 0) {
+            return false;
+        }
+
+        uint256 _tokenId = 0;
+        address _candidate;
+        bool _isMember = false;
+        while (!_isMember) {
+            _tokenId += 1;
+            _candidate = memberShipNFT.ownerOf(_tokenId);
+            _isMember = _candidate == _address;
+            console.log(_tokenId);
+            console.log(_candidate);
+        }
+        return memberShipNFT.isCommunityMember(_tokenId);
+    }
+
+    function closeQuest(uint256 _id) public {
+        require(
+            hasRole(QUEST_CREATION_ROLE, msg.sender) ||
+                hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have Quest creation role or admin role to edit a quest"
+        );
+        Quest memory _quest = quests[_id];
+        _quest.endedAt = block.timestamp;
+        _quest.endedAt = block.timestamp;
+        quests[_id] = _quest;
+    }
+
+    function update(uint256 _id, Quest memory _quest) public {
+        require(
+            hasRole(QUEST_CREATION_ROLE, msg.sender) ||
+                hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have Quest creation role or admin role to edit a quest"
+        );
+        quests[_id] = _quest;
+    }
 
     function save(Quest memory _quest) public {
-        quests.push(_quest);
+        require(
+            hasRole(QUEST_CREATION_ROLE, msg.sender) ||
+                hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have Quest creation role or admin role to add a quest"
+        );
+        quests[id] = _quest;
+        id += 1;
     }
 
-    function getQuests() public view returns(Quest[] memory){
-        return quests;
+    function canReadQuest(address _address) public view returns (bool) {
+        return
+            hasRole(QUEST_CREATION_ROLE, _address) ||
+            hasRole(ADMIN_ROLE, _address) ||
+            hasRole(HENKAKU_MEMBER_ROLE, _address) ||
+            isMember(_address);
+    }
+
+    function getQuests() public view returns (Quest[] memory) {
+        require(canReadQuest(msg.sender), "You dont have permission to obtain");
+        Quest[] memory _quests = new Quest[](id);
+        for (uint256 i = 0; i < id; i++) {
+            _quests[i] = quests[i];
+        }
+
+        return _quests;
```

Comment:
What's the difference as compared to the previous implementation of just passing `quests`?

Reply:
I changed quests data structure from Array to Hashmap for easy maintenance. e.g update, closeQuest.
but still wanted to keep same return value in `getQuests`